architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,793 @@
+```
+
+- **Important Note:** As of April 2025, Postgres is not fully implemented yet as the data store of Radius. We use etcd in production.
+
+### Architecture Diagram
+
+```mermaid
+graph TD
+    User([User]) --> CLI["Radius CLI"]
+    CLI -->|"Initiates Upgrade"| K8sAPI["Kubernetes API"]
+
+    K8sAPI --> PreflightChecks["Preflight Checks"]
+    PreflightChecks --> UserDataBackup["Backup Creation"]
+    UserDataBackup --> ComponentUpgrade["Component Upgrade"]
+    ComponentUpgrade --> PostUpgradeVerify["Post-Upgrade Verification"]
+    ComponentUpgrade -- "Failure" --> UserDataRestore["User Data Restore"]
+
+    subgraph "Radius Control Plane"
+        UCP["Universal Control Plane"]
+        DE["Deployment Engine"]
+        ARP["Applications RP"]
+        DynamicRP["Dynamic RP"]
+        Controller["Controller"]
+        Dashboard["Dashboard"]
+    end
+
+    ComponentUpgrade --> UCP
+    ComponentUpgrade --> DE
+    ComponentUpgrade --> ARP
+    ComponentUpgrade --> DynamicRP
+    ComponentUpgrade --> Controller
+    ComponentUpgrade --> Dashboard
+
+    UCP -->|"Reads/Writes"| Database["etcd / PostgreSQL"]
+    UCP -.->|"Only holds metadata"| UserApps["User Applications"]
+```
+
+### Detailed Design
+
+```mermaid
+graph TD
+    Start[User runs rad upgrade kubernetes] --> ParseArgs[Parse command arguments]
+    ParseArgs --> ValidateVersion[Validate version compatibility]
+    ValidateVersion --> AcquireLock[Acquire upgrade lock]
+    AcquireLock --> RunPreflights[Run pre-flight checks]
+    RunPreflights --> UserDataBackup[Create user data backup]
+    UserDataBackup --> PlanUpgrade[Calculate upgrade plan]
+    PlanUpgrade --> ExecuteHelmUpgrade[Execute Helm chart upgrade]
+    ExecuteHelmUpgrade --> MonitorProgress[Monitor upgrade progress]
+    MonitorProgress --> VerifyComponents[Verify component health]
+    VerifyComponents --> Success{Successful?}
+    Success -- Yes --> ReleaseLock[Release upgrade lock]
+    Success -- No --> UserDataRestore[Restore user data]
+    UserDataRestore --> RollbackHelm[Rollback Helm changes]
+    RollbackHelm --> ReleaseLock
+    ReleaseLock --> End[Display results to user]
+```
+
+**Version Detection and Validation:**
+
+The upgrade process begins with detecting the currently installed Radius version (the versions of the components in the cluster) and validating it against the requested target version. We can build an interface (or use/improve an existing one) to achieve this:
+
+```go
+// VersionValidator interface
+type VersionValidator interface {
+    ValidateTargetVersion(currentVersion, targetVersion string) error
+    GetLatestVersion() (string, error)
+    IsValidVersion(version string) bool
+}
+```
+
+This interface will be implemented (or existing will be improved) to handle version comparisons, prevent downgrades, and resolve the "latest" version tag to a specific version number. The implementation will (probably) connect to the GitHub API to fetch available release versions when needed.
```

Comment:
We need a way to check the version in an airgapped environment without the GitHub API. Maybe use the currently configured OCI container repo?

Reply:
We can list the Helm releases and get the deployed one from that.